keps/sig-apps/3939-allow-recreation-when-fully-terminated/README.md

```diff
@@ -0,0 +1,774 @@
+This is a rare case but it can provide problems if a job needs to guarantee that the existing pods terminate
+before starting new pods.  
+
+In scarce compute environments, these resources can be difficult to obtain so pods can take a long time to find resources and they may only be able to find nodes once the existing pods have been terminated.
+
+In projects that rely on tracking resource usage (like Kueue), there is a request to track the number of terminating pods as terminating pods can still use resources.
+
+If a job is stuck in terminating, it could be possible for autoscaling to kick in and give a new node.  
+This is not ideal if you could just wait for the pod to terminating and reuse that node.  
+
+### Goals
+
+- Job controller should allow for flexibility in waiting for pods to be fully terminated before
+  creating new ones
+- Job controller will have a new status field where we include the number of terminating pods.
+
+### Non-Goals
+
+- Other workload APIs are not included in this proposal.
+
+## Proposal
+
+The Job controller gets a list of active pods.  Active pods usually mean pods that have not been registered for deletion.  
+In this KEP, we will consider terminating pods to be separate from active and failed.  
+In the job controller, we should include a field that states the number of terminating pods.  
+
+We propose two new API fields:
+
+1) A field in Spec that allows for opt-in behavior of whether to wait for terminating pods to finish before recreating.
+2) A new field in Status for tracking the number of terminating pods.
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a machine learning user, ML frameworks allow scheduling of multiple pods.  
+The Job controller does not typically wait for terminating pods to be marked as failed.  
+Tensorflow and other ML frameworks may have a requirement that they only want Pods to be started once the other pods are fully terminated.
+
+This case was added due to a bug discovered with running IndexedJobs with Tensorflow.  
+See [Jobs create replacement Pods as soon as a Pod is marked for deletion](https://github.com/kubernetes/kubernetes/issues/115844) for more details.
+
+#### Story 2
+
+As a cloud user, users would want to guarantee that the number of pods that are running is exactly the amount that they specify.  
+Terminating pods do not relinguish resources so scarce compute resource are still scheduled to those pods.
+
+#### Story 3
+
+The use case in Kueue is that we want to be able to track how many pods are still terminating, regardless of whether the user wants to control the amount of running pods.  Kueue wants to track the number of pods using resources and terminating pods still utilize resources.  Without having a way to track terminating pods, it is not possible to know if a terminating pod is still using resources.
+
+See [Kueue: Account for terminating pods when doing preemption](https://github.com/kubernetes-sigs/kueue/issues/510) for an example of this.
+
```

Comment:
Not sure if this deserves a story, but I think debugging a failed job can be less confusing with this feature. In particular, it is possible that some pods that eventually terminate as Success, are incrementing the `Failed` counter. I guess this might be confusing for some users trying to understand what happened. WDYT? @alculquicondor @soltysh

Reply:
How does this feature help in that case?

Reply:
The feature actually fixes the problem, because it awaits with counting as Failed or Succeeded, until it is really terminal. 

It may not be a story, but worth mentioning somewhere. OTOH it could be a story for debugging, as the number of successful or failed pods not agreeing with the Job counter might be misleading.